### DISCUSSION TITLE
When is it legal to delete a terminated periodic sync

### DISCUSSION THREAD

**Discussion by dberlin** [zephyrproject-rtos/zephyr#69426] [Q&A]
Hey folks,
The docs are silent on when it is legal to call bt_le_per_adv_sync_delete.

In particular, is it legal to call it in the sync termination callback?
If not, at what point after is it legal to call?

If it is legal to delete in the termination callback, it would be nice to document that.
If it's not, it seems hard to decide when to delete it.

Staring at the code, it actually looks like sync flags are cleared prior to callback (it calls the internal delete function), and nothing *currently* uses it after the callback, but if that is supposed to be an API guarantee it would be nice to document it.

**Comment by SpekGit** [comment]


**Reply by OperatorHiKE** [reply]
?

**Reply by dberlin** [reply]
I don't disagree, but this is a very specific question, not a general one, so this answer is ... less helpful than I suspect you intend.

For reference: I've been doing async embedded programming for 20+ years - i'm well aware of the pitfalls here.

1. The docs here say "Can be called regardless of the state of the sync."  It's unclear if that includes "within the callback function"

2. Let's say you want to delete these when the sync is terminated, or soon after.  There is no way to look up the state of the sync  for a given object -  the struct is opaque and this state not part of the API.
As a result, the only way you know when a sync is terminated (or established) is the callback.

If you track the state yourself, and mark it as terminated in the callback for later delete by your app, you are now both duplicating state that the BT stack already has (since the state exists, it's just internal), and can easily create race conditions and sync issues within your app by duplicating that state.   That seems bad.  It is irrelevant that you can apparently delete it at any time in the sync lifecycle, this is still IMHO, a bad plan.  Especially if "regardless of the state of the sync" still requires you only do it at certain points (IE not callbacks, or at least, not *this* callback), but those points are not documented.
 
For further reference - in almost all other stacks, the syncing state is accessible.

(and, depending on the answer to this specific question, i'm happy to send a patch to zephyr to either update the docs, or add an API that lets you tell the state of a sync)

**Reply by dberlin** [reply]
Since humans don't write like this, i'm going to assume this is a bot pretending to be a human, and somebody's non-disclosed experiment.

If anyone real wants to answer, happy to talk about it.
